### ISSUE TITLE
Causing audio distortion making it impossible to 

### ISSUE THREAD

**Issue by MoonPhoenix83** [kyleneideck/BackgroundMusic#445]
**Description of the bug**
The audio is distorted and impossible to hear when my blutooth headphones are reconnected

**Steps to reproduce**
Play any audio throught background music after reconnecting bluetooth audio device (sounds like a Dalek and completely inaubible)

**Versions**
 - Background Music: 0.4.0-SNAPSHOT-c0ab98b
 - macOS 11.2 Public Beta Big Sur

**Hardware**
 - Computer: MacBook Air (Retina, 13-inch, 2020)
 - Audio Device: Sony WH-XB900N

**Comment by MoonPhoenix83** [comment]
**Update:** paused all video playing but the distortion is gone. this bug makes the app impossible to use so I have downgraded as a temporary solution

**Comment by kyleneideck** [comment]
A lot of Bluetooth devices will switch to a single channel mode in some cases. From the sound of it, that's probably not what's happening here, but it's worth checking anyway. Background Music currently only supports 2-channel devices.

**Comment by MoonPhoenix83** [comment]
By channels do you mean left/right audio so stereo?

**Comment by bloomkd46** [comment]
I have the same problem where it sounds really glitchy, it only happens when I'm using my AirPods.
Mac OS Big Sur

**Comment by kyleneideck** [comment]
@MoonPhoenix83 That's right. You can check with Audio MIDI Setup.app.

**Comment by bloomkd46** [comment]
never mind  I found out that zoom was changing my audio settings to mono.

**Comment by MoonPhoenix83** [comment]
Right, I checked the channels in Audio MIDI setup and they're all 2 channel. But, as a temporary fix, after a few tries between switching between the headphones and Background music, it worked. I will try to fiddle with zoom to find a long term solution, but this is good for an interim.

**Comment by OEverything** [comment]
When I have this problem with wired headphones, if I leave Background Music open, switch my system audio device to the computer speakers, switch the Background Music output device to the computer speakers, then select my system audio device to Background Music then change Background Music's output device to the headphones, it fixes it. Don't know if this works with Bluetooth headphones.

**Comment by OEverything** [comment]
But if that doesn't work then just mess with the hz output in Audio MIDI Setup a bit then bring it back to 48k.

**Comment by OEverything** [comment]
And if that still doesn't work then idk about a fix for it.

**Comment by iranzoferri** [comment]
Same problem here. Google meet and youtube at the same time, "meet" sounds well, but youtube only emits an inaudible noise.

**My hardware:**
- Computer: MacBook PRO (13-inch, 2020)
- Audio Device: FiiO BTR5

**Comment by OEverything** [comment]
I found that restarting the computer works to fix it. After restarting, open Background Music before you play any media and it should be fine.

**Comment by EccentricHyphus37** [comment]
## TL;DR — worth looking into codecs?

I’ve experienced the distortion too.
I’ve been wondering if this is a mid-playback codec-switch thing, though…

AFAIK, most **bluetooth two-way audio devices** (like my Sony WH-100XM3 headset) use one bluetooth profile and codec when the mic is in use ([HFP](https://en.wikipedia.org/wiki/List_of_Bluetooth_profiles#Hands-Free_Profile_(HFP)) with SCO in my case), and a much higher-quality codec when only *playing* audio ([A2DP](https://en.wikipedia.org/wiki/List_of_Bluetooth_profiles#Advanced_Audio_Distribution_Profile_(A2DP)) with AAC or SBC).
Most of the times I experience this **distortion**, my headset was communicating over **one** bluetooth profile (and therefore a corresponding audio codec), and switched to another 
While in a Microsoft Teams meeting with my headset’s microphone in use (when the bluetooth profile selected is HFP, and the codec used for audio *output* is SCO)
<img width="497" alt="image" src="https://user-images.githubusercontent.com/69630953/116709936-d2454e00-a9ee-11eb-9bab-8c27e665022f.png">

But while just *playing* audio, like on a website or local app, the A2DP profile is selected, and (depending on the bluetooth device, something like SBC or AAC):
<img width="497" alt="image" src="https://user-images.githubusercontent.com/69630953/116712225-4254d380-a9f1-11eb-872d-1323dde836ed.png">

And in my experience, it’s these *switches* that are the culprit. I’ve once successfully obtained undistorted audio through the SCO codec (with my mic on during a Teams meeting), and AAC output works fine for me most of the time.
To make it all work in either profile, I’ve had to play a lot of acrobatics, restarting the concerned audio-providing service (Teams, Music, etc.), restarting Background Music, toggling the output to the Internal Speakers in the Background Music menu and then back to my bluetooth device. And so far, only a random combination of these stunts makes it all work in a certain codec without distortion, so I tend to keep it in A2DP by refraining from using my headset mic altogether nowadays xD

Hope this helps!

**Comment by EccentricHyphus37** [comment]
Lol 🤣 I was so convinced
Now I think I was being stupid
Maybe it’s the BGM audio device?
Wondering cuz restarting `coreaudiod` seems to help me sometimes…

**Comment by cepsong** [comment]
I confirm this is a bug and it shows very often (like a buffer memory overflow issue). The only way to resolve it is to reset coreaudiod. I hope there can be an auto buffer clear function integrated to avoid such an issue from arising.
